domain/annotation/state/state.go

```diff
@@ -0,0 +1,292 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/core/annotations"
+	"github.com/juju/juju/core/database"
+	"github.com/juju/juju/domain"
+)
+
+type Annotation struct {
+	Key   string `db:"key"`
+	Value string `db:"value"`
+}
+
+// State represents a type for interacting with the underlying state.
+type State struct {
+	*domain.StateBase
+}
+
+// NewState returns a new State for interacting with the underlying state.
+func NewState(factory database.TxnRunnerFactory) *State {
+	return &State{
+		StateBase: domain.NewStateBase(factory),
+	}
+}
+
+// GetAnnotations will retrieve all the annotations associated with the given ID from the database.
+// If no annotations are found, an empty map is returned.
+func (st *State) GetAnnotations(ctx context.Context, ID annotations.ID) (map[string]string, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	kind := ID.Kind
+
+	// Prepare query for getting the UUID of ID
+	// No need if kind is model, because we keep annotations per model in the DB.
+	var kindQueryStmt *sqlair.Statement
+	var kindQuery string
+	var kindQueryParam sqlair.M
+
+	if kind != annotations.KindModel {
+		kindQuery, kindQueryParam, err = uuidQueryForID(ID)
+		if err != nil {
+			return nil, errors.Annotatef(err, "preparing get annotations query for ID: %q", ID.Name)
+		}
+		kindQueryStmt, err = sqlair.Prepare(kindQuery, sqlair.M{})
+		if err != nil {
+			return nil, errors.Annotatef(err, "preparing get annotations query for ID: %q", ID.Name)
+		}
+	}
+
+	// Prepare query for getting the annotations of ID
+	var getAnnotationsQuery string
+	var getAnnotationsStmt *sqlair.Statement
+
+	if kind == annotations.KindModel {
+		getAnnotationsQuery = `SELECT (key, value) AS (&Annotation.*) from annotation_model`
+	} else {
+		getAnnotationsQuery = fmt.Sprintf(`
+SELECT (key, value) AS (&Annotation.*)
+FROM annotation_%[1]s
+WHERE %[1]s_uuid = $M.uuid
+`, kind)
```

Comment:
I'd rather we passed those variables in via `?`. So construct them outside of the query and then pass them in during the query phase. Otherwise, it's rife for SQL injection.